cannon/mipsevm/multithreaded/mips.go

```diff
@@ -343,36 +343,40 @@ func (m *InstrumentedState) handleRMWOps(insn, opcode uint32) error {
 	base := m.state.GetRegistersRef()[baseReg]
 	rtReg := Word((insn >> 16) & 0x1F)
 	offset := exec.SignExtendImmediate(insn)
+	addr := base + offset
 
-	effAddr := (base + offset) & arch.AddressMask
-	m.memoryTracker.TrackMemAccess(effAddr)
-	mem := m.state.Memory.GetWord(effAddr)
+	// Determine some opcode-specific parameters
+	targetStatus := LLStatusActive32bit
+	byteLength := Word(4)
+	if opcode == exec.OpLoadLinked64 || opcode == exec.OpStoreConditional64 {
+		// Use 64-bit params
+		targetStatus = LLStatusActive64bit
+		byteLength = Word(8)
+	}
 
 	var retVal Word
 	threadId := m.state.GetCurrentThread().ThreadId
-	if opcode == exec.OpLoadLinked || opcode == exec.OpLoadLinked64 {
-		retVal = mem
-		m.state.LLReservationActive = true
-		m.state.LLAddress = effAddr
+	switch opcode {
+	case exec.OpLoadLinked, exec.OpLoadLinked64:
+		retVal = exec.LoadSubWord(m.state.GetMemory(), addr, byteLength, true, m.memoryTracker)
+
+		m.state.LLReservationStatus = targetStatus
+		m.state.LLAddress = addr
 		m.state.LLOwnerThread = threadId
-	} else if opcode == exec.OpStoreConditional || opcode == exec.OpStoreConditional64 {
-		// TODO(#12205): Determine bits affected by coherence stores on 64-bits
-		// Check if our memory reservation is still intact
-		if m.state.LLReservationActive && m.state.LLOwnerThread == threadId && m.state.LLAddress == effAddr {
+	case exec.OpStoreConditional, exec.OpStoreConditional64:
+		if m.state.LLReservationStatus == targetStatus && m.state.LLOwnerThread == threadId && m.state.LLAddress == addr {
```

Comment:
The store is now conditional on the virtual address. I think we should use the effective address instead because the `LLAddress` is cleared based on that.
Otherwise, if `LLAddress=0x101` and a regular store to 0x101 occurred, then we `sc[d]` will erroneously succeed. This scenario may be contrived though, since the vaddr should already be naturally aligned (to 4 and 8 bytes for sc and scd) and I don't expect compilers violate this alignment constraint. Nevertheless, it's good to be extra defensive here.

Reply:
We now mask the `LLAddress` with `arch.AddressMask` [here](https://github.com/ethereum-optimism/optimism/blob/de438d3ad389b61603dac01e230e232f77b29e42/cannon/mipsevm/multithreaded/mips.go#L328) when checking if a reservation should be invalidated.  So I don't think we'll fail to invalidate unless I'm missing something?  We could defensively mask both addresses when checking, but the `memAddr` value there should already be masked.

The reserved address is a little complicated in 64-bit MIPS.  For the `ll`/`sc` instructions, we want to make sure the 4-byte address offset matches between load and store.  But if we just store the effective address to `LLAddress`, that offset bit will be lost.

Or are you suggesting we mask the LLAddress based on the operation type?  So, for LL/SC, we store a 4-byte aligned value and for LLD/SCD, we store an 8-byte aligned value?  I think that could work, but it seems to complicate the code without giving us much benefit since we need to mask it when we do the memory check anyway. 

There's also this in the MIPS `sc` documentation, which implies we should be doing stricter checks:
> A RMW sequence executed without intervening exceptions must use the same address in the LL and SC.The address is the same if the virtual address, physical address, and cache-coherence algorithm are identical.

Reply:
> Or are you suggesting we mask the LLAddress based on the operation type? So, for LL/SC, we store a 4-byte aligned value and for LLD/SCD, we store an 8-byte aligned value? I think that could work, but it seems to complicate the code without giving us much benefit since we need to mask it when we do the memory check anyway.

I think so because the current mask breaks single-word ll/sc semantics. Consider when there is a RMW sequence initiated for a lower word (ex: 0x104) and a 32-bit instruction issues a write to the bits between [0x100, 0x104). A `sc` would succeed on a 32-bit CPU, so I'd expect the same for the 64-bit CPU per backwards compatibility. However, `handleMemoryUpdate` clears the memory reserved at 0x104 since it's located on the same effective address.

Reply:
This might be OK. The spec has this to say about such stores:
> A coherent store is completed by another processor or coherent I/O module into the block of synchronizable
physical memory containing the word. The size and alignment of the block is implementation-dependent, but it is at least one word and at most the minimum page size.

So technically, cannon64 would be within spec. It's just that it results spurious sc failures if a store occurs within the same doubelword block. In which case, I'd optimize for implementation simplicity like you suggested rather than try to eliminate these spurious sc failures.